certora/specs/SocialRecoveryModule.spec

```diff
@@ -257,3 +275,26 @@ rule confirmRecoveryCanAlwaysBeInitiatedByGuardian(env e, address guardian, addr
         currentContract.recoveryRequests[safeContract].newOwners.length == newOwners.length &&
         currentContract.recoveryRequests[safeContract].newOwners[index] == newOwners[index];
 }
+
+// This rule verifies that the finalization cannot happen if the recovery module is not enabled.
+// Exceptions are made for the case where the Safe has only one owner and the recovery is initiated
+// with zero new owners and zero as the new threshold.
+rule disabledRecoveryModuleResultsInFinalizationRevert(env e) {
+    address[] currentOwners = safeContract.getOwners();
+    uint256 currentThreshold = safeContract.getThreshold();
+
+    require !safeContract.isModuleEnabled(currentContract);
+
+    currentContract.finalizeRecovery@withrevert(e, safeContract);
+    bool finalizeRecoveryIsReverted = lastReverted;
+
+    // If the recovery finalization is initiated with the safe having only one owner,
+    // and the finalize recovery initiated with no new owners and zero as new threshold,
+    // then the finalize recovery call goes through, as no owner is removed and no new
+    // owner is added. Though it is not possible to have a recovery initiation with zero
+    // owners.
+    assert finalizeRecoveryIsReverted ||
+        (currentOwners[0] == safeContract.getOwners()[0] &&
+            safeContract.getOwners().length == 1 &&
+            currentThreshold == safeContract.getThreshold());
```

Comment:
This should not be needed this way.

In particular, we have this check in the recovery module:

https://github.com/5afe/CandideWalletContracts/blob/d218bdd2259e5d2282230ab1b463db8f05383970/contracts/modules/social_recovery/SocialRecoveryModule.sol#L160

This means that we should be able to _prove_ that newOwners is never an empty list and never run into this edge case.

Otherwise, I worry we may be hiding bugs.

Concretely:
- Write an invariant that proves that you are never in the `newOwners = [] and newThreshold = 0` case
- `requireInvariant` that into this rule

Reply:
Even with that invariant, this check might be required because of this:

https://github.com/5afe/CandideWalletContracts/blob/d218bdd2259e5d2282230ab1b463db8f05383970/contracts/modules/social_recovery/SocialRecoveryModule.sol#L251

The `newOwner` can be the same owner as the safe's current owner, with just one owner for that safe.